### PR TITLE
Random add rangeReal()

### PR DESCRIPTION
**Pull request by breznak**

### CODE REVIEW

File: src/nupic/utils/Random.hpp [hunk 1]
```diff
@@ -122,13 +122,6 @@ class Random : public Serializable  {
     pop.resize(nChoices); //keep only first nChoices, drop rest
     return pop;
   }
-  //compatibility method for Py-bindings, //TODO remove with SWIG
```
Comment (breznak) [inline src/nupic/utils/Random.hpp]: I'm removing old wrapper for Random::sample, 
updating Sdr.cpp and py bindings.

File: bindings/py/cpp_src/bindings/math/py_Random.cpp [hunk 2]
```diff
@@ -63,16 +63,10 @@ namespace nupic_ext {
 
 
         Random.def("sample",
-            [](Random_t& r, py::array_t<nupic::UInt32>& population, py::array_t<nupic::UInt32>& choices)
+            [](Random_t& r, std::vector<nupic::UInt32>& population, nupic::UInt32 nChoices)
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/math/py_Random.cpp]: @dkeeney why are we using py::array (-> numpy.array) and not std::vector (-> list[] in py) ? 
- This compiles, implementation is simpler
  - I could just adapt tests to pass

- can pybind even automatically wrap functions where possible? (so we wouldn't have to write this dummy wrapper)
  - from py, if you have numpy.array, you can easily get list, not the other way around. 

Do you agree to this modification (and possibly prefered way of writing the wrappers)?

File: src/nupic/utils/Random.hpp [hunk 3]
```diff
@@ -122,14 +122,18 @@ class Random : public Serializable  {
...
-  void sample(const T population[], UInt nPopulation, T choices[], UInt nChoices) {
-    std::vector<T> vPop(population, population + nPopulation); 
-    std::vector<T> vChoices = this->sample<T>(vPop, nChoices);
-    std::copy(vChoices.begin(), vChoices.end(), choices);
+
+
+  /**
+   * return random from range [from, to)
+   */
+  Real realRange(Real from, Real to) {
```
Comment (breznak) [inline src/nupic/utils/Random.hpp]: this is the new method - used ie to initialize synapses

File: bindings/py/cpp_src/bindings/math/py_Random.cpp (hunk 2 above)
Comment (breznak) [reply bindings/py/cpp_src/bindings/math/py_Random.cpp]: true, thanks! 
I'll revert this change for this PR. 
For discussion if we can find a simpler method than py::array: https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html

File: bindings/py/cpp_src/bindings/math/py_Random.cpp [hunk 4]
```diff
@@ -63,16 +63,15 @@ namespace nupic_ext {
...
+            if (population.ndim() != 1 )
             {
                 throw std::runtime_error("Number of dimensions must be one.");
             }
 
-            r.sample(get_it(population), (nupic::UInt)population.size(), get_it(choices), (nupic::UInt)choices.size());
-
-            return choices;
+	    std::vector<nupic::UInt32> tmp_pop(get_it(population), get_it(population) + population.size()); //vector from numpy.array
+            return r.sample(tmp_pop, nSelect);
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/math/py_Random.cpp]: I'm returning py list, and should return py::array_t , how do I do that?

Comment (ctrl-z-9000-times) [reply bindings/py/cpp_src/bindings/math/py_Random.cpp]: The pybind11 docs have a section on numpy arrays, which is worth a read:
https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html

Or you can look at the SDR bindings which are in the algorithms directory.  Look for `py::array`.  Here is an example from the SDR bindings: (don't ask me what the capsule does)
```
auto capsule = py::capsule(&self, [](void *self) {});
return py::array(self.getSparse().size(), self.getSparse().data(), capsule);
```

**Comment by breznak** [comment]
Thanks for reviewing! 

> Also, I'm not worried about the whole python-list versus numpy-array issue, until a profiler tells me that it's an issue. 

This is a good approach, imho. The list makes the pybind extentions functionality even easier to use, we might consider that for future.

**Pull request merged**